### PR TITLE
Update # issues returned by num_label_issues

### PR DESCRIPTION
**Pull request by ulya-tkch**
- [x] code
- [x] tests
- [x] docstring
- [nit] If num_label_issues is called with passed in calibrated ``confident_joint`` matrix but ``estimation_method='off_diagonal'`` should we throw a warning suggesting the ``confident_joint`` might be already calibrated or let it be?

### CODE REVIEW

File: cleanlab/count.py [hunk 1]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
...
+       two cases:
+          1. As we add more label and data quality scoring functions in :py:mod:`cleanlab.rank`, this approach will always work.
+          2. If you have a custom score to rank your data by label quality and you just need to know the cut-off of likely label issues.
+
     Returns
     -------
     num_issues : int
       The estimated number of examples with label issues in the dataset.
     """
+    valid_methods = ["off_diagonal", "off_diagonal_recalibrated"]
```
[the remark this answers is not in the document]
Comment (cgnorthcutt) [reply cleanlab/count.py:105]: (or better is add static types in the method declaration)

File: cleanlab/count.py [hunk 2]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
...
-    num_issues = np.rint(frac_issues * len(labels)).astype(int)
+        confident_joint = compute_confident_joint(
+            labels=labels, pred_probs=pred_probs, calibrate=False
+        )
+
+    if estimation_method is "off_diagonal":
+        num_issues = np.sum(confident_joint) - np.trace(confident_joint)
+    elif estimation_method is "off_diagonal_recalibrated":
+        # Normalize confident joint so that it estimates the joint, p(labels,y)
+        confident_joint = calibrate_confident_joint(confident_joint, labels)
```
Comment (cgnorthcutt) [inline cleanlab/count.py:116]:
```suggestion
```

Comment (cgnorthcutt) [reply cleanlab/count.py:116]: delete this line

File: cleanlab/count.py [hunk 3]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
...
     """
+    valid_methods = ["off_diagonal", "off_diagonal_recalibrated"]
 
     if confident_joint is None:
-        confident_joint = compute_confident_joint(labels=labels, pred_probs=pred_probs)
-    # Normalize confident joint so that it estimates the joint, p(labels,y)
-    joint = confident_joint / float(np.sum(confident_joint))
-    frac_issues = 1.0 - joint.trace()
-    num_issues = np.rint(frac_issues * len(labels)).astype(int)
+        confident_joint = compute_confident_joint(
```
Comment (cgnorthcutt) [inline cleanlab/count.py:108]: Add a comment above `# Original non-calibrated counts of confidently correctly and incorrectly labeled examples.`

File: cleanlab/count.py [hunk 4]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
...
+            labels=labels, pred_probs=pred_probs, calibrate=False
+        )
+
+    if estimation_method is "off_diagonal":
+        num_issues = np.sum(confident_joint) - np.trace(confident_joint)
+    elif estimation_method is "off_diagonal_recalibrated":
+        # Normalize confident joint so that it estimates the joint, p(labels,y)
+        confident_joint = calibrate_confident_joint(confident_joint, labels)
+        joint = confident_joint / float(np.sum(confident_joint))
```
Comment (cgnorthcutt) [inline cleanlab/count.py:117]:
```suggestion
        joint = estimate_joint(labels, pred_probs, confident_joint=confident_joint)
```

this is the cleanlab-way to do it in one line of code. achieves identical result as the previous code just uses the actual function (which it should)

File: tests/test_filter_count.py [hunk 5]
```diff
@@ -427,14 +427,64 @@ def test_find_label_issue_filters_match_origin_functions():
...
+        labels=data["labels"],
+        pred_probs=data["pred_probs"],
+        confident_joint=data["cj"],
+        estimation_method="off_diagonal_recalibrated",
+    )  # data["cj"] is already calibrated but recalibrating it should not change the values
+
+    n2 = count.num_label_issues(
+        labels=data["labels"],
+        pred_probs=data["pred_probs"],
+        estimation_method="off_diagonal_recalibrated",
```
Comment (cgnorthcutt) [inline tests/test_filter_count.py:446]:
```suggestion
        estimation_method="off_diagonal_calibrated",
```

File: tests/test_filter_count.py [hunk 6]
```diff
@@ -427,14 +427,64 @@ def test_find_label_issue_filters_match_origin_functions():
...
+        pred_probs=data["pred_probs"],
+        confident_joint=data["cj"],
+        estimation_method="off_diagonal",
+    )  # data["cj"] is already calibrated and estimation method does not do extra calibration
+
+    n1 = count.num_label_issues(
+        labels=data["labels"],
+        pred_probs=data["pred_probs"],
+        confident_joint=data["cj"],
+        estimation_method="off_diagonal_recalibrated",
```
Comment (cgnorthcutt) [inline tests/test_filter_count.py:440]:
```suggestion
        estimation_method="off_diagonal_calibrated",
```

File: cleanlab/count.py [hunk 7]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
       The `confident_joint` can be computed using :py:func:`count.compute_confident_joint <cleanlab.count.compute_confident_joint>`.
       If not provided, it is computed from the given (noisy) `labels` and `pred_probs`.
 
+    estimation_method :
+      Method for estimating the number of label issues in dataset by counting the examples in the off-diagonal of the `confident_joint` ``P(label=i, true_label=j)``.
+       - ``'off_diagonal'``: Counts the number of examples in the off-diagonal of the `confident_joint`. Returns the same value as ``sum(find_label_issues(filter_by='confident_learning'))``
+       - ``'off_diagonal_recalibrated'``: Calibrates confident joint estimate ``P(label=i, true_label=j)`` such that
```
[the remark this answers is not in the document]
Comment (cgnorthcutt) [reply cleanlab/count.py:91]: its only calibrated here for the first time. just say calibrated, not recalibrated

File: cleanlab/count.py (hunk 4 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:117]: Note that `confident_joint` is not calibrated but this method calibrates it and then normalizes (makes it sum to 1) which is what the line above did.

you could add a comment above this line that says:

`# estimate_joint calibrates the row sums to match the prior distribution of given labels and normalizes to sum to 1`

**Review by cgnorthcutt — CHANGES_REQUESTED** [review]
Overall looks good. Some minor changes. 

change `recalibrated` to `calibrated` everywhere (its only calibrated the one time)

One decision (up to you if you want to leave this off for the PR, but its definitely worth benchmarking and potentially updating):
 * for the `off_diagonals` method, we should still calibrate the rows to correctly sum to the true prior (p(s)) of noisy given labels (without also normalizing to sum to the total number of examples) -- this would make the rows more accurately affect the prior distribution without overcounting errors. Should run the 9x4 benchmark examples to make sure the counts are reasonable (and more accurate) if you add this update. (fyi - @jwmueller )

File: cleanlab/count.py [hunk 8]
```diff
@@ -96,18 +85,45 @@ class 0, 1, ..., K-1. `pred_probs` should have been computed using 3 (or
       The `confident_joint` can be computed using :py:func:`count.compute_confident_joint <cleanlab.count.compute_confident_joint>`.
       If not provided, it is computed from the given (noisy) `labels` and `pred_probs`.
 
+    estimation_method :
+      Method for estimating the number of label issues in dataset by counting the examples in the off-diagonal of the `confident_joint` ``P(label=i, true_label=j)``.
```
Comment (jwmueller) [inline cleanlab/count.py:89]: @cgnorthcutt Could you elaborate on this:
> for the off_diagonals method, we should still calibrate the rows to correctly sum to the true prior (p(s)) of noisy given labels (without also normalizing to sum to the total number of examples) -- this would make the rows more accurately affect the prior distribution without overcounting errors. 

The main point of this PR is to make the num_label_issues count match find_label_issues when off_diagonal is the method. Wouldn't any change like you are suggesting make them no longer match?

File: cleanlab/count.py [hunk 9]
```diff
@@ -56,23 +56,12 @@
...
-
-    This method is **more accurate** than ``sum(find_label_issues())`` because its computed using only
-    the trace of the confident joint, ignoring all off-diagonals (which are used by `find_label_issues` and are harder to
-    estimate). Here, we sum over only diagonal elements in the joint (which have more data
-    are more constrained, and therefore easier to compute).
-
-    TL;DR: use this method to get the most accurate estimate of number of label issues when you
-    don't need the indices of the label issues.
-
-    You can use this method to label issues by using `num_label_issues` as the cutoff threshold
```
[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/count.py:71]: @ulya-tkch You could move some of these descriptions under the docstring describing the argument ``'off_diagonal_recalibrated'``

File: cleanlab/count.py (hunk 8 above)
Comment (cgnorthcutt) [reply cleanlab/count.py:89]: i would still benchmark to see if its more accurate and if it is, then it sure another `method` added to this function called `off_diagonal_row_calibrated`

Initial benchmarks suggest this `off_diagonal` method is more accurate (in addition to matching `find_label_issues(filter_by='confident_learning')`) but may slightly undercount.

File: cleanlab/count.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (ulya-tkch) [reply cleanlab/count.py:71]: I omitted because was unsure if we should call it the "most accurate estimate" after the benchmarking. Will move back in under `off_diagonal_recalibrated`

[the remark this answers is not in the document]
Comment (jwmueller) [reply cleanlab/count.py:71]: ya let's not call it most accurate, just say it is another way to estimate that adjusts for under-estimation bias in the other method which limits itself to only count the confident subset of examples

**Comment by ulya-tkch** [comment]
Closes #371

**Review by jwmueller — APPROVED** [review]

**Review by cgnorthcutt — APPROVED** [review]
LGTM. great PR @ulya-tkch
